### ISSUE TITLE
ActivePeers on Node 8.0.0 (P2P-Mode) getting lower over time

### ISSUE THREAD

**Issue by gitmachtl** [input-output-hk/ouroboros-network#4563]
Upgraded normally from cardano-node 1.35.7 to cardano-node 8.0.0 on one of my Mainnet-Relais.

The P2P-Settings in the config.json are:
``` json
"EnableP2P": true,
  "TargetNumberOfActivePeers": 30,
  "TargetNumberOfEstablishedPeers": 50,
  "TargetNumberOfKnownPeers": 100,
  "TargetNumberOfRootPeers": 100,
  "TraceAcceptPolicy": true,
  "TraceInboundGovernor": true,
  "TraceLedgerPeers": true,
  "TraceLocalRootPeers": true,
  "TraceConnectionManager": true,
```

After a fresh start, the node gets 30 active peers, all seems normal. But after a few minutes and over the following hours, the number of active peers is reduced one by one. Sent some logs to @coot and @karknu, looks like the node did not start to demote the peers in case an active one gets decreased.

From the logs:
```
2023-05-16T06:09:05.20Z Normal Churn starts
2023-05-16T06:09:05.20Z targetActivePeers is decreased from 30 to 24
2023-05-16T06:09:06.20Z targetActivePeers increased to 30 , no demotion of hot peers has started
2023-05-16T06:09:07.21Z targetEstablishedPeers is decreased to from 50 to 40
                        targetKnownPeers is decreased from 100 to 80
            targetRootPeers is decreased from 100 to 80
2023-05-16T06:09:07.21Z 20 warm peers are picked for demotion, actualEstablished=100 and targetEstablished=40
2023-05-16T06:11:08.20Z Churn Done
2023-05-16T06:11:08.21Z targetEstablished is increased to 50
                        targetKnownPeers is increased to 100
            targetRootPeers is increased to 100
2023-05-16T06:11:08.21Z 20 cold peers are picked for promotion
```

> Karl: The governor has 1s to detect the decrease in targetActivePeers and start to demote them.
> It is apparently busy with other things so the churn doesn't happen.

The machine is a decent one with 12 cores, also did some tests with only the node running. Its chillin all the time.
Test was done with a self-compiled cardano-node binary and also with the pre-built binary from the 8.0.0 release.

Will repeat a test with only 20 targetActivePeers and also Karl sent me a special commit to try out. Such an issue did not happen with node 1.35.7. For a test i reverted the node back to 1.35.7 and the connections were stable. 

Best regards,
 Martin

**Comment by gitmachtl** [comment]
These are the stat after around 13,5 hours of uptime with the special commit from @karknu. The activePeers are down again from 30 to 24, will let it run to collect more log data:

``` js
$ curl -s http://127.0.0.1:12798/metrics | grep cardano_node | sort
cardano_node_metrics_blockfetchclient_blockdelay_cdfFive 1.0
cardano_node_metrics_blockfetchclient_blockdelay_cdfOne 0.9768518518518519
cardano_node_metrics_blockfetchclient_blockdelay_cdfThree 0.9953703703703703
cardano_node_metrics_blockfetchclient_blockdelay_s 0.181514782
cardano_node_metrics_blockfetchclient_blocksize 40694
cardano_node_metrics_blockfetchclient_lateblocks 9
cardano_node_metrics_blockNum_int 8796812
cardano_node_metrics_connectionManager_duplexConns 53
cardano_node_metrics_connectionManager_incomingConns 145
cardano_node_metrics_connectionManager_outgoingConns 112
cardano_node_metrics_connectionManager_prunableConns 0
cardano_node_metrics_connectionManager_unidirectionalConns 204
cardano_node_metrics_density_real 4.980630879911455e-2
cardano_node_metrics_epoch_int 412
cardano_node_metrics_forks_int 28
cardano_node_metrics_inboundGovernor_cold 3
cardano_node_metrics_inboundGovernor_hot 95
cardano_node_metrics_inboundGovernor_idle 0
cardano_node_metrics_inboundGovernor_warm 50
cardano_node_metrics_Mem_resident_int 14846513152
cardano_node_metrics_nodeStartTime_int 1684519786
cardano_node_metrics_peers_connectedPeers_int 24
cardano_node_metrics_peerSelection_cold 3
cardano_node_metrics_peerSelection_hot 24
cardano_node_metrics_peerSelection_warm 88
cardano_node_metrics_RTS_gcHeapBytes_int 14754512896
cardano_node_metrics_RTS_gcLiveBytes_int 9341410632
cardano_node_metrics_RTS_gcMajorNum_int 35
cardano_node_metrics_RTS_gcMinorNum_int 33936
cardano_node_metrics_RTS_gcticks_int 350887
cardano_node_metrics_RTS_mutticks_int 283682
cardano_node_metrics_served_block_count_int 19699
cardano_node_metrics_served_block_latest_count_int 17784
cardano_node_metrics_served_header_counter_int 389076
cardano_node_metrics_slotInEpoch_int 381032
cardano_node_metrics_slotNum_int 93001832
cardano_node_metrics_Stat_cputicks_int 634570
cardano_node_metrics_Stat_threads_int 24
```

Will do another test with the release version of the node and a TargetNumberOfActivePeers set to 20. However, i have already 9 localRoots, so only 11 external slots left than.

**Comment by gitmachtl** [comment]
ok, what i did:
- pulled a fresh database snapshot from csnapshots.io
- removed the entries for
```json
"TargetNumberOfActivePeers": 30,
"TargetNumberOfEstablishedPeers": 50,
"TargetNumberOfKnownPeers": 100,
"TargetNumberOfRootPeers": 100,
```
from the config.json
- removed the 2nd section in the localRoots in the topology.json (the leftover are just 9 entries for the localRoots with a valency set to 9)

current state after around 5hours uptime, 20 peers active like it should with the release version of node 8.0.0.

i will now start to reintroduce the config.json paramters and also re-add the 2nd section in the localRoots topology.json. something must make the node angry.

**Comment by coot** [comment]
@karknu's patch which might help: https://github.com/karknu/ouroboros-network/commit/63179fe18388bf71cc8142928b493d075dd08c50

**Comment by gitmachtl** [comment]
@coot @karknu So, as pointed out above, the "special Karl patch" did not help. BUT, i think i found the problem.

I reintroduced the parameters
``` json
"TargetNumberOfActivePeers": 30,
"TargetNumberOfEstablishedPeers": 50,
"TargetNumberOfKnownPeers": 100,
"TargetNumberOfRootPeers": 100,
```
into the config.json and restarted the node. After 12 hours of uptime it was stable with 30 peers active.

Only thing left to test was the topology file, and there we go. If i use a "normal" p2p topology file in the format:
``` json
{
  "localRoots": [
      { "accessPoints": [
                { "name": "xxxx", "address": "x.x.x.x", "port": 3001 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3002 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3003 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3004 },
                { "name": "yyyy", "address": "y.y.y.y", "port": 3005 },
                { "name": "Local-Backup-1", "address": "127.0.0.1", "port": 1001 },
                { "name": "Local-Backup-2", "address": "127.0.0.1", "port": 1002 },
                { "name": "Local-Backup-3", "address": "127.0.0.1", "port": 1003 },
                { "name": "Local-Backup-4", "address": "127.0.0.1", "port": 1004 }
        ],
        "advertise": false,
        "valency": 9
      }
  ],
  "publicRoots": [
    { "accessPoints": [
                { "address": "relays-new.cardano-mainnet.iohk.io", "port": 3001 }
        ],
      "advertise": false
    }
  ],
  "useLedgerAfterSlot": 85366187
}
``` 
all is good. Stable active peers after hours of uptime.

BUT when i introduce a second part into the localRoots section like:
``` json
{
  "localRoots": [
      { "accessPoints": [
                { "name": "xxxx", "address": "x.x.x.x", "port": 3001 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3002 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3003 },
                { "name": "xxxx", "address": "x.x.x.x", "port": 3004 },
                { "name": "yyyy", "address": "y.y.y.y", "port": 3005 },
                { "name": "Local-Backup-1", "address": "127.0.0.1", "port": 1001 },
                { "name": "Local-Backup-2", "address": "127.0.0.1", "port": 1002 },
                { "name": "Local-Backup-3", "address": "127.0.0.1", "port": 1003 },
                { "name": "Local-Backup-4", "address": "127.0.0.1", "port": 1004 }
        ],
        "advertise": false,
        "valency": 9 },
      { "accessPoints": [
                { "name": "zzzz", "address": "z.z.z.z", "port": 1234 },
                { "address": "relays-new.cardano-mainnet.iohk.io", "port": 3001 }
        ],
        "advertise": false,
        "valency": 3 }
  ],
  "publicRoots": [
    { "accessPoints": [
                { "address": "relays-new.cardano-mainnet.iohk.io", "port": 3001 }
        ],
      "advertise": false
    }
  ],
  "useLedgerAfterSlot": 85366187
}
```
it falls apart and the cardano-node is doing that weired peer degredation stuff. Valenvy=3 was set in the 2nd part so the node keeps one connection to zzzz (only resolves to one ip) and two connection to the mainnet-relays too.

If i run it without the 2nd part, all is good. So, did i mess it up or is there a problem reading the topology file correctly?

**Comment by coot** [comment]
@gitmachtl well done! We now know where to look for.  Will the behaviour persist if you change the domain name in the added local roots to `IP` addresses?

**Comment by gitmachtl** [comment]
> @gitmachtl well done! We now know where to look for. Will the behaviour persist if you change the domain name in the added local roots to `IP` addresses?

I replaced the `zzzz` entry with the ip address and i replaced the `relays-new.cardano-mainnet.iohk.io` now by two ip-address entries and restarted the node. Lets see...

**Comment by coot** [comment]
btw, it seems you are running at least 5 nodes on a single machine and you are likely connecting to another 4 nodes which are running on a single host.  What's the CPU/memory usage on these nodes? Are there any spikes when we drop connections?

**Comment by gitmachtl** [comment]
No spikes at all, these are all machines with 64-128GB of ram and between 12 and 20 cores. And nodes are not all running. They are just in the topology for a backup situation.

**Comment by gitmachtl** [comment]
Short update: Using only ip addresses for the 2nd part in the localRoots worked, its stable with 30 active peers. 

I am now rerunning the test with valency=3 but only using two ip address entries.

**Comment by gitmachtl** [comment]
Update: The node is now running for around 3,5 hours without an active peer count degredation.

So, there must be something strange when it comes into combination with host-name resolving.

**Comment by coot** [comment]
@gitmachtl Do you have `TraceLocalRootPeers` and `TracePublicRootPeers` tracers enabled?

**Comment by gitmachtl** [comment]
> @gitmachtl Do you have `TraceLocalRootPeers` and `TracePublicRootPeers` tracers enabled?

`TraceLocalRootPeers` is enabled, `TracePublicRootPeers` is not set. I can do a rerun with both enabled if you want me too for some logs?

**Comment by coot** [comment]
The first one should be enough in this case (I would turn the other on the next restart).

**Comment by bolt12** [comment]
How long does it take for you to observe the issue? What if you only have 1 DNS name in the group (hopefully not the 1 that only resolves to 1 IP)?

**Comment by gitmachtl** [comment]
I sent @coot again a full log today, he is also testing it currently. Have not reduced to only one DNS. But with the settings above and the two FQDN entries, it took only 2 hours to degrade from 30 to 23 peers.

**Comment by gitmachtl** [comment]
@bolt12 i am still wondering why node 1.35.7 is not showing this behaviour to fill of the localRoots with multiple different ip addresses from a FQDN entry?

**Comment by bolt12** [comment]
@gitmachtl, 1.35.7 had an issue related to DNS, as outlined in https://github.com/input-output-hk/ouroboros-network/issues/4491. This led to a situation where the DNS resolver would overwrite the results of other DNS lookups when multiple DNS domains were in use within the same group, effectively prohibiting the appropriate regulation of IPs in local roots.

We addressed this issue in version 8.0.0 but failed to consider that #4565 could arise, notice that this issue would still exist in the case when the user configures hundreds of local root peers using IPs. This suggests that addressing this inherent issue at its core, rather than applying a superficial fix, would be a more prudent approach.

Our initial plan is to tackle the problem stemming from how DNS resolution could lead to this situation. We greatly appreciate your patience and understanding as we work towards implementing this solution! 😄

**Comment by bolt12** [comment]
Not to mention all the help and effort you've been putting to help us debug this issue! Thank you!

**Comment by gitmachtl** [comment]
> Not to mention all the help and effort you've been putting to help us debug this issue! Thank you!

you're welcome

**Comment by coot** [comment]
I am testing [this branch](https://github.com/input-output-hk/cardano-node/commits/coot/local-roots-fix) of `cardano-node` (forked of `8.0.0`), which includes the fix. @gitmachtl could you also check if it fixes the issue for you.

**Comment by gitmachtl** [comment]
@coot compiling right now... will let you know how it goes.

**Comment by coot** [comment]
It seems to work for me:
```
root@coot:/data/cardano-node# jq 'select(.data.kind=="LocalRootPeersChanged") | .data.current.groups | .[1] | .[1] | length' ./logs/cardano-node.log -c
8
16
16
16
...
```
and the node keept 30 hot peers after a churn:
```
root@coot:/data/cardano-node# jq -c 'select(.data.kind == "PeerSelectionCounters") | {"at":.at, "hot": .data.hotPeers}'  ./logs/cardano-node.log | tail
{"at":"2023-05-26T09:40:46.54Z","hot":30}
{"at":"2023-05-26T09:40:56.44Z","hot":30}
{"at":"2023-05-26T09:41:10.76Z","hot":30}
{"at":"2023-05-26T09:41:12.81Z","hot":30}
{"at":"2023-05-26T09:41:29.20Z","hot":30}
{"at":"2023-05-26T09:41:46.59Z","hot":30}
{"at":"2023-05-26T09:41:55.84Z","hot":30}
{"at":"2023-05-26T09:42:14.25Z","hot":30}
{"at":"2023-05-26T09:42:46.65Z","hot":30}
{"at":"2023-05-26T09:42:57.29Z","hot":30}
```
Churn happened at:
```
root@coot:/data/cardano-node# jq -c 'select(.data.kind == "ChurnMode") | {"at":.at, "mode": .data.event}'  ./logs/cardano-node.log
{"at":"2023-05-26T08:54:28.29Z","mode":"ChurnModeNormal"}
```

**Comment by gitmachtl** [comment]
jumped from 30 to 27 here after 1hour of uptime. lets see how it progresses...
ok, few minutes later, its back at 30 hot peers.

**Comment by bolt12** [comment]
Let's give the nodes some more time to stabilize 😄 But promising preliminary observations!

**Comment by bolt12** [comment]
Github closed this because the fix got merged. Let's only close it when we get a confirmation that it works

**Comment by coot** [comment]
My node survived three churn periods for me:
```
{"at":"2023-05-26T08:54:28.29Z","mode":"ChurnModeNormal"}
{"at":"2023-05-26T09:54:55.38Z","mode":"ChurnModeNormal"}
{"at":"2023-05-26T10:54:28.74Z","mode":"ChurnModeNormal"}
```
Everytime recovering back to `30` hot peers.

**Comment by coot** [comment]
@gitmachtl I think we can close it, if you still encounter this please re-open.  There's still a chance to be hit by the bug if a domain name resolves to too many addresses, but this will not be fixed in `8.1.0` (see #4565).
